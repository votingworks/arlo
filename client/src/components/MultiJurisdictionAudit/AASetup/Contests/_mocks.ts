import { IContest } from '../../../../types'

export const contestMocks: {
  [key: string]: { contests: IContest[] }
} = {
  emptyTargeted: {
    contests: [
      {
        id: '',
        name: '',
        isTargeted: true,
        totalBallotsCast: '',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: [],
        choices: [
          {
            id: '',
            name: '',
            numVotes: '',
          },
          {
            id: '',
            name: '',
            numVotes: '',
          },
        ],
      },
    ],
  },
  emptyOpportunistic: {
    contests: [
      {
        id: '',
        name: '',
        isTargeted: false,
        totalBallotsCast: '',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: [],
        choices: [
          {
            id: '',
            name: '',
            numVotes: '',
          },
          {
            id: '',
            name: '',
            numVotes: '',
          },
        ],
      },
    ],
  },
  filledTargeted: {
    contests: [
      {
        id: 'contest-id',
        name: 'Contest Name',
        isTargeted: true,
        totalBallotsCast: '30',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: [],
        choices: [
          {
            id: 'choice-id-1',
            name: 'Choice One',
            numVotes: '10',
          },
          {
            id: 'choice-id-2',
            name: 'Choice Two',
            numVotes: '20',
          },
        ],
      },
    ],
  },
  filledOpportunistic: {
    contests: [
      {
        id: 'contest-id',
        name: 'Contest Name',
        isTargeted: false,
        totalBallotsCast: '30',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: [],
        choices: [
          {
            id: 'choice-id-3',
            name: 'Choice Three',
            numVotes: '10',
          },
          {
            id: 'choice-id-4',
            name: 'Choice Four',
            numVotes: '20',
          },
        ],
      },
    ],
  },
  filledTargetedWithJurisdictionId: {
    contests: [
      {
        id: 'contest-id',
        name: 'Contest Name',
        isTargeted: true,
        totalBallotsCast: '30',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: ['jurisdiction-id-1', 'jurisdiction-id-2'],
        choices: [
          {
            id: 'choice-id-1',
            name: 'Choice One',
            numVotes: '10',
          },
          {
            id: 'choice-id-2',
            name: 'Choice Two',
            numVotes: '20',
          },
        ],
      },
    ],
  },
  filledOpportunisticWithJurisdictionId: {
    contests: [
      {
        id: 'contest-id',
        name: 'Contest Name',
        isTargeted: false,
        totalBallotsCast: '30',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: ['jurisdiction-id-1', 'jurisdiction-id-2'],
        choices: [
          {
            id: 'choice-id-3',
            name: 'Choice Three',
            numVotes: '10',
          },
          {
            id: 'choice-id-4',
            name: 'Choice Four',
            numVotes: '20',
          },
        ],
      },
    ],
  },
  filledTargetedAndOpportunistic: {
    contests: [
      {
        id: 'contest-id',
        name: 'Contest 1',
        isTargeted: true,
        totalBallotsCast: '30',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: ['jurisdiction-id-1', 'jurisdiction-id-2'],
        choices: [
          {
            id: 'choice-id-1',
            name: 'Choice One',
            numVotes: '10',
          },
          {
            id: 'choice-id-2',
            name: 'Choice Two',
            numVotes: '20',
          },
        ],
      },
      {
        id: 'contest-id',
        name: 'Contest 2',
        isTargeted: false,
        totalBallotsCast: '30',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: ['jurisdiction-id-1', 'jurisdiction-id-2'],
        choices: [
          {
            id: 'choice-id-3',
            name: 'Choice Three',
            numVotes: '10',
          },
          {
            id: 'choice-id-4',
            name: 'Choice Four',
            numVotes: '20',
          },
        ],
      },
    ],
  },
}

export const contestsInputMocks = {
  inputs: [
    { key: 'Contest Name', value: 'Contest Name' },
    { key: 'Name of Candidate/Choice 1', value: 'Choice One' },
    { key: 'Name of Candidate/Choice 2', value: 'Choice Two' },
    { key: 'Votes for Candidate/Choice 1', value: '10' },
    { key: 'Votes for Candidate/Choice 2', value: '20' },
    { key: 'Total Ballots for Contest', value: '30' },
  ],
  errorInputs: [
    { key: 'Contest Name', value: '', error: 'Required' },
    {
      key: 'Total Ballots for Contest',
      value: '',
      error:
        'Must be greater than or equal to the sum of votes for each candidate/choice',
    },
    {
      key: 'Total Ballots for Contest',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Total Ballots for Contest',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Total Ballots for Contest',
      value: '0.5',
      error: 'Must be an integer',
    },
    { key: 'Name of Candidate/Choice 1', value: '', error: 'Required' },
    { key: 'Name of Candidate/Choice 2', value: '', error: 'Required' },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '',
      error: 'Required',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '0.5',
      error: 'Must be an integer',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '',
      error: 'Required',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '0.5',
      error: 'Must be an integer',
    },
  ],
}

export default contestsInputMocks
