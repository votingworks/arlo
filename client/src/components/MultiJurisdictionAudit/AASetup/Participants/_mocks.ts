import { readFileSync } from 'fs'
import { join } from 'path'

const jurisdictionFile = new File(
  [
    readFileSync(
      join(
        __dirname,
        '../../../../../public/sample_jurisdiction_filesheet.csv'
      ),
      'utf8'
    ),
  ],
  'jurisdictions.csv',
  { type: 'text/csv' }
)

export default jurisdictionFile
