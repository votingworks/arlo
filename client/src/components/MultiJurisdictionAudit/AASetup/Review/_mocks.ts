import { IAuditSettings } from '../../../../types'

export const settingsMock: {
  [key in 'empty' | 'full' | 'offline' | 'batch']: IAuditSettings
} = {
  empty: {
    state: null,
    electionName: null,
    online: null,
    randomSeed: null,
    riskLimit: null,
    auditType: 'BALLOT_POLLING',
    auditName: 'Test Audit',
  },
  full: {
    state: 'AL',
    electionName: 'Election Name',
    online: true,
    randomSeed: '12345',
    riskLimit: 10,
    auditType: 'BALLOT_POLLING',
    auditName: 'Test Audit',
  },
  offline: {
    state: 'AL',
    electionName: 'Election Name',
    online: false,
    randomSeed: '12345',
    riskLimit: 10,
    auditType: 'BALLOT_POLLING',
    auditName: 'Test Audit',
  },
  batch: {
    state: 'AL',
    electionName: 'Election Name',
    online: false,
    randomSeed: '12345',
    riskLimit: 10,
    auditType: 'BATCH_COMPARISON',
    auditName: 'Test Audit',
  },
}

export const sampleSizeMock = {
  sampleSizes: {
    'contest-id': [
      { prob: 0.54, size: 46, key: 'asn' },
      { prob: 0.7, size: 67, key: '0.7' },
      { prob: 0.5, size: 88, key: '0.5' },
      { prob: 0.9, size: 125, key: '0.9' },
    ],
  },
}

export default settingsMock
