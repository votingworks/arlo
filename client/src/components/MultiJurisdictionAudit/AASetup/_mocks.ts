import { ElementType } from '../../../types'
import { setupStages, stageTitles } from './index'
import { ISidebarMenuItem } from '../../Atoms/Sidebar'

const relativeStages = (
  stage: ElementType<typeof setupStages>,
  state: ISidebarMenuItem['state'] = 'live'
): {
  prevStage: ISidebarMenuItem
  nextStage: ISidebarMenuItem
  menuItems: ISidebarMenuItem[]
} => {
  const prevId = setupStages[setupStages.indexOf(stage) - 1]
  const prevStage: ISidebarMenuItem = {
    id: prevId,
    title: stageTitles[prevId],
    active: false,
    activate: jest.fn(),
    state,
  }
  const nextId = setupStages[setupStages.indexOf(stage) + 1]
  const nextStage: ISidebarMenuItem = {
    id: nextId,
    title: stageTitles[nextId],
    active: false,
    activate: jest.fn(),
    state,
  }
  const menuItems = setupStages.map((s: ElementType<typeof setupStages>) => ({
    id: s,
    title: stageTitles[s],
    active: s === stage,
    activate: jest.fn(),
    state,
  }))
  return { prevStage, nextStage, menuItems }
}

export default relativeStages
