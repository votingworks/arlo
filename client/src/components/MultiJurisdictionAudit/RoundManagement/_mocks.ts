export interface INullResultValues {
  [contestId: string]: {
    [choiceId: string]: null | string | number
  }
}

export const resultsMocks: {
  [key in 'emptyInitial' | 'complete']: INullResultValues
} = {
  emptyInitial: {
    'contest-id-1': {
      'choice-id-1': null,
      'choice-id-2': null,
    },
    'contest-id-2': {
      'choice-id-3': null,
      'choice-id-4': null,
    },
  },
  complete: {
    'contest-id-1': {
      'choice-id-1': 1,
      'choice-id-2': 2,
    },
    'contest-id-2': {
      'choice-id-3': 1,
      'choice-id-4': 2,
    },
  },
}

export default resultsMocks
