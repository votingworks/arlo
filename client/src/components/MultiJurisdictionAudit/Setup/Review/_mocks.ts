export const settingsMock = {
  empty: [
    {
      state: null,
      electionName: null,
      online: null,
      randomSeed: null,
      riskLimit: null,
      auditType: 'BALLOT_POLLING',
    },
    // here for type completion but not used in this context
    /* istanbul ignore next */
    async () => true,
  ],
  full: [
    {
      state: 'AL',
      electionName: 'Election Name',
      online: true,
      randomSeed: '12345',
      riskLimit: 10,
      auditType: 'BALLOT_POLLING',
    },
    // here for type completion but not used in this context
    /* istanbul ignore next */
    async () => true,
  ],
  offline: [
    {
      state: 'AL',
      electionName: 'Election Name',
      online: false,
      randomSeed: '12345',
      riskLimit: 10,
      auditType: 'BALLOT_POLLING',
    },
    // here for type completion but not used in this context
    /* istanbul ignore next */
    async () => true,
  ],
}

export const sampleSizeMock = {
  sampleSizes: {
    'contest-id': [
      { prob: 0.54, size: 46, key: 'asn' },
      { prob: 0.7, size: 67, key: '0.7' },
      { prob: 0.5, size: 88, key: '0.5' },
      { prob: 0.9, size: 125, key: '0.9' },
    ],
  },
}

export default settingsMock
