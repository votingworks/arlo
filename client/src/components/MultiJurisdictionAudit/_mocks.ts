import { contestMocks } from './AASetup/Contests/_mocks'
import { IFileInfo } from './useJurisdictions'
import {
  auditSettings,
  manifestFile,
  talliesFile,
} from './useSetupMenuItems/_mocks'
import { FileProcessingStatus } from './useSetupMenuItems/getJurisdictionFileStatus'
import { IAuditSettings } from '../../types'
import { IBallot } from './RoundManagement/useBallots'
import { IAuditBoard } from './useAuditBoards'
import { IRound } from './useRoundsAuditAdmin'

const manifestFormData: FormData = new FormData()
manifestFormData.append('manifest', manifestFile, manifestFile.name)
const talliesFormData: FormData = new FormData()
talliesFormData.append('batchTallies', talliesFile, talliesFile.name)

export const jaApiCalls = {
  getUser: {
    url: '/api/me',
    response: {
      type: 'jurisdiction_admin',
      name: 'Joe',
      email: 'test@email.org',
      jurisdictions: [
        {
          id: 'jurisdiction-id-1',
          name: 'Jurisdiction One',
          election: {
            id: '1',
            auditName: 'audit one',
            electionName: 'election one',
            state: 'AL',
          },
        },
        {
          id: 'jurisdiction-id-2',
          name: 'Jurisdiction Two',
          election: {
            id: '2',
            auditName: 'audit two',
            electionName: 'election two',
            state: 'AL',
          },
        },
        {
          id: 'jurisdiction-id-3',
          name: 'Jurisdiction Three',
          election: {
            id: '1',
            auditName: 'audit one',
            electionName: 'election one',
            state: 'AL',
          },
        },
      ],
      organizations: [],
    },
  },
  getRounds: {
    url: '/api/election/1/jurisdiction/jurisdiction-id-1/round',
    response: { rounds: [] },
  },
  getBallotManifestFile: (response: IFileInfo) => ({
    url: '/api/election/1/jurisdiction/jurisdiction-id-1/ballot-manifest',
    response,
  }),
  getBatchTalliesFile: (response: IFileInfo) => ({
    url: '/api/election/1/jurisdiction/jurisdiction-id-1/batch-tallies',
    response,
  }),
  getCVRSfile: (response: IFileInfo) => ({
    url: '/api/election/1/jurisdiction/jurisdiction-id-1/cvrs',
    response,
  }),
  getSettings: (response: IAuditSettings) => ({
    url: '/api/election/1/jurisdiction/jurisdiction-id-1/settings',
    response,
  }),
  putManifest: {
    url: '/api/election/1/jurisdiction/jurisdiction-id-1/ballot-manifest',
    options: {
      method: 'PUT',
      body: manifestFormData,
    },
    response: { status: 'ok' },
  },
  putTallies: {
    url: '/api/election/1/jurisdiction/jurisdiction-id-1/batch-tallies',
    options: {
      method: 'PUT',
      body: talliesFormData,
    },
    response: { status: 'ok' },
  },
  getAuditBoards: (auditBoards: IAuditBoard[]) => ({
    url:
      '/api/election/1/jurisdiction/jurisdiction-id-1/round/round-1/audit-board',
    response: { auditBoards },
  }),
  getBallots: (ballots: IBallot[]) => ({
    url: '/api/election/1/jurisdiction/jurisdiction-id-1/round/round-1/ballots',
    response: { ballots },
  }),
}

export const aaApiCalls = {
  getUser: {
    url: '/api/me',
    response: {
      type: 'audit_admin',
      name: 'Joe',
      email: 'test@email.org',
      jurisdictions: [],
      organizations: [
        {
          id: 'org-id',
          name: 'State of California',
          elections: [],
        },
      ],
    },
  },
  getRounds: (rounds: IRound[]) => ({
    url: '/api/election/1/round',
    response: { rounds },
  }),
  getJurisdictions: {
    url: '/api/election/1/jurisdiction',
    response: {
      jurisdictions: [
        {
          id: 'jurisdiction-id-1',
          name: 'Jurisdiction One',
          ballotManifest: { file: null, processing: null },
          currentRoundStatus: null,
        },
        {
          id: 'jurisdiction-id-2',
          name: 'Jurisdiction Two',
          ballotManifest: { file: null, processing: null },
          currentRoundStatus: null,
        },
      ],
    },
  },
  getBatchJurisdictions: {
    url: '/api/election/1/jurisdiction',
    response: {
      jurisdictions: [
        {
          id: 'jurisdiction-id-1',
          name: 'Jurisdiction One',
          ballotManifest: { file: null, processing: null },
          batchTallies: { file: null, processing: null },
          currentRoundStatus: null,
        },
        {
          id: 'jurisdiction-id-2',
          name: 'Jurisdiction Two',
          ballotManifest: { file: null, processing: null },
          batchTallies: { file: null, processing: null },
          currentRoundStatus: null,
        },
      ],
    },
  },
  getJurisdictionFile: {
    url: '/api/election/1/jurisdiction/file',
    response: {
      file: {
        contents: null,
        name: 'file name',
        uploadedAt: 'a long time ago in a galaxy far far away',
      },
      processing: {
        status: FileProcessingStatus.Processed,
        error: null,
        startedAt: 'once upon a time',
        endedAt: 'and they lived happily ever after',
      },
    },
  },
  getContests: {
    url: '/api/election/1/contest',
    response: contestMocks.filledTargeted,
  },
  getSettings: (response: IAuditSettings) => ({
    url: '/api/election/1/settings',
    response,
  }),
  putSettings: {
    url: '/api/election/1/settings',
    options: {
      method: 'PUT',
      body: JSON.stringify(auditSettings.all),
      headers: { 'Content-Type': 'application/json' },
    },
    response: { status: 'ok' },
  },
  getSampleSizes: {
    url: '/api/election/1/sample-sizes',
    response: { sampleSizes: null },
  },
}
