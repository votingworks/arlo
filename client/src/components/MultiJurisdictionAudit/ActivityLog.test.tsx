import React from 'react'
import { screen, waitFor } from '@testing-library/react'
import { QueryClientProvider } from 'react-query'
import userEvent from '@testing-library/user-event'
import { aaApiCalls } from './_mocks'
import { withMockFetch, renderWithRouter } from '../testUtilities'
import ActivityLog from './ActivityLog'
import AuthDataProvider from '../UserContext'
import { PrivateRoute, queryClient } from '../../App'
import * as utilities from '../utilities'

const mockElection = {
  id: 'election-id-1',
  auditName: 'Test Audit',
  auditType: 'BALLOT_POLLING',
}

const mockAuditAdmin = {
  key: 'admin@example.gov',
  type: 'audit_admin',
  supportUser: null,
}

const mockJurisdictionAdmin = {
  key: 'admin@example.gov',
  type: 'jurisdiction_admin',
  supportUser: null,
}

const mockSupportUser = {
  key: 'admin@example.gov',
  type: 'audit_admin',
  supportUser: 'support@example.gov',
}

// Adds a minute each time it's called
const nextTimestamp = (() => {
  let lastTimestamp = new Date('2021-08-31T22:52:49.762Z')
  return () => {
    lastTimestamp = new Date(lastTimestamp.getTime() + 60 * 1000)
    return lastTimestamp
  }
})()

const nextId = (() => {
  let lastId = 0
  return () => {
    lastId += 1
    return lastId
  }
})()

const apiCalls = {
  getActivities: {
    url: `/api/organizations/org-id/activities`,
    response: [
      {
        id: nextId(),
        activityName: 'CreateAudit',
        timestamp: nextTimestamp(),
        election: mockElection,
        user: mockAuditAdmin,
        info: {},
      },
      {
        id: nextId(),
        activityName: 'JurisdictionAdminLogin',
        timestamp: nextTimestamp(),
        election: mockElection,
        user: mockJurisdictionAdmin,
        info: { error: 'Invalid code' },
      },
      {
        id: nextId(),
        activityName: 'JurisdictionAdminLogin',
        timestamp: nextTimestamp(),
        election: mockElection,
        user: mockJurisdictionAdmin,
        info: { error: null },
      },
      {
        id: nextId(),
        activityName: 'UploadFile',
        timestamp: nextTimestamp(),
        election: mockElection,
        user: null,
        info: {
          jurisdiction_id: 'jurisdiction-id-1', // eslint-disable-line @typescript-eslint/camelcase
          jurisdiction_name: 'Jurisdiction 1', // eslint-disable-line @typescript-eslint/camelcase
          file_type: 'ballot_manifest', // eslint-disable-line @typescript-eslint/camelcase
          error: 'failed',
        },
      },
      {
        id: nextId(),
        activityName: 'UploadFile',
        timestamp: nextTimestamp(),
        election: mockElection,
        user: null,
        info: {
          jurisdiction_id: 'jurisdiction-id-1', // eslint-disable-line @typescript-eslint/camelcase
          jurisdiction_name: 'Jurisdiction 1', // eslint-disable-line @typescript-eslint/camelcase
          file_type: 'batch_tallies', // eslint-disable-line @typescript-eslint/camelcase
          error: null,
        },
      },
      {
        id: nextId(),
        activityName: 'UploadFile',
        timestamp: nextTimestamp(),
        election: mockElection,
        user: null,
        info: {
          jurisdiction_id: 'jurisdiction-id-1', // eslint-disable-line @typescript-eslint/camelcase
          jurisdiction_name: 'Jurisdiction 1', // eslint-disable-line @typescript-eslint/camelcase
          file_type: 'cvrs', // eslint-disable-line @typescript-eslint/camelcase
          error: null,
        },
      },
      {
        id: nextId(),
        activityName: 'CalculateSampleSizes',
        timestamp: nextTimestamp(),
        election: mockElection,
        user: mockAuditAdmin,
        info: {},
      },
      {
        id: nextId(),
        activityName: 'StartRound',
        timestamp: nextTimestamp(),
        election: mockElection,
        user: mockSupportUser,
        info: { round_num: 1 }, // eslint-disable-line @typescript-eslint/camelcase
      },
      {
        id: nextId(),
        activityName: 'CreateAuditBoards',
        timestamp: nextTimestamp(),
        election: mockElection,
        user: mockJurisdictionAdmin,
        info: {
          jurisdiction_id: 'jurisdiction-id-1', // eslint-disable-line @typescript-eslint/camelcase
          jurisdiction_name: 'Jurisdiction 1', // eslint-disable-line @typescript-eslint/camelcase
          num_audit_boards: 2, // eslint-disable-line @typescript-eslint/camelcase
        },
      },
      {
        id: nextId(),
        activityName: 'RecordResults',
        timestamp: nextTimestamp(),
        election: mockElection,
        user: mockJurisdictionAdmin,
        info: {
          jurisdiction_id: 'jurisdiction-id-1', // eslint-disable-line @typescript-eslint/camelcase
          jurisdiction_name: 'Jurisdiction 1', // eslint-disable-line @typescript-eslint/camelcase
        },
      },
      {
        id: nextId(),
        activityName: 'AuditBoardSignOff',
        timestamp: nextTimestamp(),
        election: mockElection,
        user: {
          key: 'audit-board-id-1',
          type: 'audit_board',
          supportUser: false,
        },
        info: {
          jurisdiction_id: 'jurisdiction-id-1', // eslint-disable-line @typescript-eslint/camelcase
          jurisdiction_name: 'Jurisdiction 1', // eslint-disable-line @typescript-eslint/camelcase
          audit_board_name: 'Audit Board #1', // eslint-disable-line @typescript-eslint/camelcase
        },
      },
      {
        id: nextId(),
        activityName: 'EndRound',
        timestamp: nextTimestamp(),
        election: mockElection,
        user: {
          key: 'audit-board-id-1',
          type: 'audit_board',
          supportUser: false,
        },
        info: { round_num: 1, is_audit_complete: true }, // eslint-disable-line @typescript-eslint/camelcase
      },
      {
        id: nextId(),
        activityName: 'DeleteAudit',
        timestamp: nextTimestamp(),
        election: mockElection,
        user: mockAuditAdmin,
        info: {},
      },
    ].reverse(),
  },
}

const renderActivityLog = () =>
  renderWithRouter(
    <QueryClientProvider client={queryClient}>
      <AuthDataProvider>
        <PrivateRoute userType="audit_admin" component={ActivityLog} />
      </AuthDataProvider>
    </QueryClientProvider>
  )

describe('Activity Log', () => {
  it('shows a table of activity for the org', async () => {
    const expectedCalls = [aaApiCalls.getUser, apiCalls.getActivities]
    await withMockFetch(expectedCalls, async () => {
      renderActivityLog()
      await screen.findByRole('heading', { name: 'Activity Log' })
      expect(
        screen.getAllByRole('columnheader').map(h => h.textContent)
      ).toEqual(['Timestamp', 'User', 'Action', 'Audit', 'Jurisdiction'])
      expect(screen.getByRole('table')).toMatchSnapshot()
    })
  })

  it('has a dropdown for audit admins with multiple orgs', async () => {
    const expectedCalls = [
      aaApiCalls.getUserMultipleOrgs,
      apiCalls.getActivities,
      { url: '/api/organizations/org-id-2/activities', response: [] },
    ]
    await withMockFetch(expectedCalls, async () => {
      renderActivityLog()
      await screen.findByRole('heading', { name: 'Activity Log' })
      // Loads the first org by default
      const orgSelect = screen.getByLabelText(/Organization:/)
      expect(orgSelect).toHaveTextContent('State of California')
      expect(screen.getAllByRole('row')).toHaveLength(
        apiCalls.getActivities.response.length + 1
      )

      // Select a different org
      userEvent.selectOptions(orgSelect, 'State of Georgia')
      await waitFor(() => expect(screen.getAllByRole('row')).toHaveLength(1))
    })
  })

  it('has a button to download the activity table as a CSV', async () => {
    // JSDOM doesn't implement innerText, so we implement it using textContent
    Object.defineProperty(HTMLElement.prototype, 'innerText', {
      get() {
        return this.textContent
      },
      configurable: true,
    })
    const downloadFileMock = jest
      .spyOn(utilities, 'downloadFile')
      .mockImplementation()

    const expectedCalls = [
      aaApiCalls.getUserMultipleOrgs,
      apiCalls.getActivities,
    ]
    await withMockFetch(expectedCalls, async () => {
      renderActivityLog()
      await screen.findByRole('heading', { name: 'Activity Log' })
      userEvent.click(screen.getByRole('button', { name: /Download as CSV/ }))
      expect(downloadFileMock).toHaveBeenCalled()
      expect(downloadFileMock.mock.calls[0][1]).toMatch(
        /arlo-activity-State of California/
      )
      const fileBlob = downloadFileMock.mock.calls[0][0] as Blob
      expect(fileBlob.type).toEqual('text/csv')
      expect(await new Response(fileBlob).text()).toMatchInlineSnapshot(`
        "\\"Timestamp\\",\\"User\\",\\"Action\\",\\"Audit\\",\\"Jurisdiction\\"
        \\"8/31/2021, 11:05:49 PM\\",\\"admin@example.gov\\",\\"Deleted audit\\",\\"Test Audit\\",\\"\\"
        \\"8/31/2021, 11:04:49 PM\\",\\"\\",\\"Ended round 1\\",\\"Test Audit\\",\\"\\"
        \\"8/31/2021, 11:03:49 PM\\",\\"\\",\\"Audit Board #1 signed off\\",\\"Test Audit\\",\\"Jurisdiction 1\\"
        \\"8/31/2021, 11:02:49 PM\\",\\"admin@example.gov\\",\\"Recorded results\\",\\"Test Audit\\",\\"Jurisdiction 1\\"
        \\"8/31/2021, 11:01:49 PM\\",\\"admin@example.gov\\",\\"Created audit boards\\",\\"Test Audit\\",\\"Jurisdiction 1\\"
        \\"8/31/2021, 11:00:49 PM\\",\\"support@example.gov\\",\\"Started round 1\\",\\"Test Audit\\",\\"\\"
        \\"8/31/2021, 10:59:49 PM\\",\\"admin@example.gov\\",\\"Calculated sample sizes\\",\\"Test Audit\\",\\"\\"
        \\"8/31/2021, 10:58:49 PM\\",\\"\\",\\"Successfully uploaded CVRs\\",\\"Test Audit\\",\\"Jurisdiction 1\\"
        \\"8/31/2021, 10:57:49 PM\\",\\"\\",\\"Successfully uploaded candidate totals by batch\\",\\"Test Audit\\",\\"Jurisdiction 1\\"
        \\"8/31/2021, 10:56:49 PM\\",\\"\\",\\"Uploaded invalid ballot manifest\\",\\"Test Audit\\",\\"Jurisdiction 1\\"
        \\"8/31/2021, 10:55:49 PM\\",\\"admin@example.gov\\",\\"Logged in as a Jurisdiction Manager\\",\\"Test Audit\\",\\"\\"
        \\"8/31/2021, 10:54:49 PM\\",\\"admin@example.gov\\",\\"Failed to log in as a Jurisdiction Manager: Invalid code\\",\\"Test Audit\\",\\"\\"
        \\"8/31/2021, 10:53:49 PM\\",\\"admin@example.gov\\",\\"Created audit\\",\\"Test Audit\\",\\"\\""
      `)

      downloadFileMock.mockRestore()
      delete HTMLElement.prototype.innerText
    })
  })
})
