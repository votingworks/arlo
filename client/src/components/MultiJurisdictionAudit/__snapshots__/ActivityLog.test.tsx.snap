// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity Log shows a table of activity for the org 1`] = `
<table
  class="sc-htpNat cJosdG"
  id="activityLog"
  style="table-layout: auto;"
>
  <thead>
    <tr>
      <th>
        Timestamp
      </th>
      <th>
        User
      </th>
      <th>
        Action
      </th>
      <th>
        Audit
      </th>
      <th>
        Jurisdiction
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        8/31/2021, 11:05:49 PM
      </td>
      <td>
        admin@example.gov
      </td>
      <td>
        Deleted audit
      </td>
      <td>
        Test Audit
      </td>
      <td />
    </tr>
    <tr>
      <td>
        8/31/2021, 11:04:49 PM
      </td>
      <td />
      <td>
        Ended round 1
      </td>
      <td>
        Test Audit
      </td>
      <td />
    </tr>
    <tr>
      <td>
        8/31/2021, 11:03:49 PM
      </td>
      <td />
      <td>
        Audit Board #1 signed off
      </td>
      <td>
        Test Audit
      </td>
      <td>
        Jurisdiction 1
      </td>
    </tr>
    <tr>
      <td>
        8/31/2021, 11:02:49 PM
      </td>
      <td>
        admin@example.gov
      </td>
      <td>
        Recorded results
      </td>
      <td>
        Test Audit
      </td>
      <td>
        Jurisdiction 1
      </td>
    </tr>
    <tr>
      <td>
        8/31/2021, 11:01:49 PM
      </td>
      <td>
        admin@example.gov
      </td>
      <td>
        Created audit boards
      </td>
      <td>
        Test Audit
      </td>
      <td>
        Jurisdiction 1
      </td>
    </tr>
    <tr>
      <td>
        8/31/2021, 11:00:49 PM
      </td>
      <td>
        support@example.gov
      </td>
      <td>
        Started round 1
      </td>
      <td>
        Test Audit
      </td>
      <td />
    </tr>
    <tr>
      <td>
        8/31/2021, 10:59:49 PM
      </td>
      <td>
        admin@example.gov
      </td>
      <td>
        Calculated sample sizes
      </td>
      <td>
        Test Audit
      </td>
      <td />
    </tr>
    <tr>
      <td>
        8/31/2021, 10:58:49 PM
      </td>
      <td />
      <td>
        Successfully uploaded CVRs
      </td>
      <td>
        Test Audit
      </td>
      <td>
        Jurisdiction 1
      </td>
    </tr>
    <tr>
      <td>
        8/31/2021, 10:57:49 PM
      </td>
      <td />
      <td>
        Successfully uploaded candidate totals by batch
      </td>
      <td>
        Test Audit
      </td>
      <td>
        Jurisdiction 1
      </td>
    </tr>
    <tr>
      <td>
        8/31/2021, 10:56:49 PM
      </td>
      <td />
      <td>
        Uploaded invalid ballot manifest
      </td>
      <td>
        Test Audit
      </td>
      <td>
        Jurisdiction 1
      </td>
    </tr>
    <tr>
      <td>
        8/31/2021, 10:55:49 PM
      </td>
      <td>
        admin@example.gov
      </td>
      <td>
        Logged in as a Jurisdiction Manager
      </td>
      <td>
        Test Audit
      </td>
      <td />
    </tr>
    <tr>
      <td>
        8/31/2021, 10:54:49 PM
      </td>
      <td>
        admin@example.gov
      </td>
      <td>
        <span>
          Failed to log in as a Jurisdiction Manager: 
          <br />
          Invalid code
        </span>
      </td>
      <td>
        Test Audit
      </td>
      <td />
    </tr>
    <tr>
      <td>
        8/31/2021, 10:53:49 PM
      </td>
      <td>
        admin@example.gov
      </td>
      <td>
        Created audit
      </td>
      <td>
        Test Audit
      </td>
      <td />
    </tr>
  </tbody>
</table>
`;
